pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,578 @@
+		newHostChains = append(newHostChains, hostChains...)
+		newServiceChains = append(newServiceChains, svcChain)
+
+		// write chain and sticky session rule
+		for _, hostChain := range hostChains {
+			// Create chain
+			if chain, ok := existingChains[utiliptables.Chain(hostChain)]; ok {
+				chainsLines.WriteString(fmt.Sprintf("%s\n", chain))
+			} else {
+				chainsLines.WriteString(makeChainLine(hostChain))
+			}
+
+			// Sticky session
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --rcheck --seconds %d --reap -j %s\n", svcChain, name.String(), hostChain, info.stickyMaxAgeMinutes*60, hostChain))
+			}
+		}
+
+		// write proxy/loadblanacing rules
+		n := len(hostChains)
+		for i, hostChain := range hostChains {
+			// Roughly round robin statistically if we have more than one host
+			if n > 1 {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m statistic --mode random --probability %f -j %s\n", svcChain, name.String(), 1.0/float64(n-i), hostChain))
+			} else {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j %s\n", svcChain, name.String(), hostChain))
+			}
+			// proxy
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --set -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), hostChain, protocol, hosts[i]))
+			} else {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), protocol, hosts[i]))
+			}
+		}
+
+		// proxy
+		rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), info.portal.ip.String(), protocol, protocol, info.portal.port, svcChain))
+
+		for _, publicIP := range info.deprecatedPublicIPs {
+			rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"deprecated-PublicIP portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), publicIP, protocol, protocol, info.portal.port, svcChain))
+		}
+
+		for _, ingress := range info.loadBalancerStatus.Ingress {
+			if ingress.IP != "" {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"load-balancer portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), ingress.IP, protocol, protocol, info.portal.port, svcChain))
+			}
+		}
+	}
+
+	// Delete chains no longer in use:
+	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
+	for _, chain := range newHostChains {
+		activeChains[chain] = true
+	}
+	for _, chain := range newServiceChains {
+		activeChains[chain] = true
+	}
+	for chain := range existingChains {
```

Comment:
Yeah. The recent pushes haven't been as thoroughly tested yet since firing
up a cluster is really slow for me (vagrant). I'm mostly checking for
builds and integration tests during the day and watching the e2e results
then going over the cluster stuff locally when things quiet down eg at the
end of the day.

I'm working with a cluster locally right now fixing some of the stuff
mentioned here so far.
I'll try to have everything patched tonight so we can move on tomorrow.

On Wed, Aug 12, 2015 at 1:25 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36827532>
> :
>
> > +		for _, ingress := range info.loadBalancerStatus.Ingress {
> > +			if ingress.IP != "" {
> > +				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"load-balancer portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), ingress.IP, protocol, protocol, info.portal.port, svcChain))
> > +			}
> > +		}
> > +	}
> > +
> > +	// Delete chains no longer in use:
> > +	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
> > +	for _, chain := range newHostChains {
> > +		activeChains[chain] = true
> > +	}
> > +	for _, chain := range newServiceChains {
> > +		activeChains[chain] = true
> > +	}
> > +	for chain := range existingChains {
>
> need to beef up your manual testing at least. Would be nice to cover these
> with e2e but your time is short
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36827532>
> .
>